stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,602 @@
+    }
+
+    return rootElement;
+  }
+
+  public void findMatchingElements(String query, Accumulator<Integer> matchedIds) {
+    verifyThreadAccess();
+
+    final Pattern queryPattern = Pattern.compile(Pattern.quote(query), Pattern.CASE_INSENSITIVE);
+    final Object rootElement = mDocumentProvider.getRootElement();
+
+    findMatches(queryPattern, rootElement, matchedIds);
+  }
+
+  private void findMatches(Pattern queryPattern, Object element, Accumulator<Integer> matchedIds) {
+    final ElementInfo info = mShadowDOM.getElementInfo(element);
+
+    for (int i = 0, size = info.children.size(); i < size; i++) {
+      final Object childElement = info.children.get(i);
+
+      if (doesElementMatch(queryPattern, childElement)) {
+        matchedIds.store(mObjectIdMapper.getIdForObject(childElement));
+      }
+
+      findMatches(queryPattern, childElement, matchedIds);
+    }
+  }
+
+  private boolean doesElementMatch(Pattern queryPattern, Object element) {
+    AttributeListAccumulator accumulator = new AttributeListAccumulator();
+    NodeDescriptor descriptor = mDocumentProvider.getNodeDescriptor(element);
+
+    descriptor.getAttributes(element, accumulator);
+
+    for (int i = 0, N = accumulator.size(); i < N; i++) {
+      if (queryPattern.matcher(accumulator.get(i)).find()) {
+        return true;
+      }
+    }
+
+    return queryPattern.matcher(descriptor.getNodeName(element)).find();
+  }
+
+  private ChildEventingList acquireChildEventingList(
+      Object parentElement,
+      DocumentView documentView) {
+    ChildEventingList childEventingList = mCachedChildEventingList;
+    if (childEventingList == null) {
+      childEventingList = new ChildEventingList();
+    }
+    childEventingList.acquire(parentElement, documentView);
+    return childEventingList;
+  }
+
+  private void releaseChildEventingList(ChildEventingList childEventingList) {
+    childEventingList.release();
+    if (mCachedChildEventingList == null) {
+      mCachedChildEventingList = childEventingList;
+    }
+  }
+
+  private ArrayListAccumulator<Object> acquireChildrenAccumulator() {
+    ArrayListAccumulator<Object> accumulator = mCachedChildrenAccumulator;
+    if (accumulator == null) {
+      accumulator = new ArrayListAccumulator<>();
+    }
+    mCachedChildrenAccumulator = null;
+    return accumulator;
+  }
+
+  private void releaseChildrenAccumulator(ArrayListAccumulator<Object> accumulator) {
+    accumulator.clear();
+    if (mCachedChildrenAccumulator == null) {
+      mCachedChildrenAccumulator = accumulator;
+    }
+  }
+
+  private ShadowDocument.Update createShadowDOMUpdate() {
+    verifyThreadAccess();
+
+    if (mDocumentProvider.getRootElement() != mShadowDOM.getRootElement()) {
+      throw new IllegalStateException();
+    }
+
+    ArrayListAccumulator<Object> childrenAccumulator = acquireChildrenAccumulator();
+
+    ShadowDocument.UpdateBuilder updateBuilder = mShadowDOM.beginUpdate();
+    Queue<Object> queue = new ArrayDeque<>();
```

Comment:
I think we can add a `// TODO: cache this` on here too, and then resolve this before we're done with this PR.